project_wbs/models/project_project.py

```diff
@@ -0,0 +1,264 @@
+            data = '/'.join(data)
+            res2 = project_item.code_get()
+            if res2 and res2[0][1]:
+                data = '[' + res2[0][1] + '] ' + data
+
+            res.append((project_item.id, data))
+        return res
+
+    @api.multi
+    @api.depends('code')
+    def code_get(self):
+        res = []
+        for project_item in self:
+            data = []
+            proj = project_item
+            while proj:
+                if proj.code:
+                    data.insert(0, proj.code)
+                else:
+                    data.insert(0, '')
+
+                proj = proj.parent_id
+
+            data = '/'.join(data)
+            res.append((project_item.id, data))
+        return res
+
+    @api.multi
+    @api.depends('parent_id')
+    def _compute_child(self):
+        for project_item in self:
+            child_ids = self.search(
+                [('parent_id', '=', project_item.analytic_account_id.id)]
+            )
+            project_item.project_child_complete_ids = child_ids
+
+    @api.multi
+    def _resolve_analytic_account_id_from_context(self):
+        """
+        Returns ID of parent analytic account based on the value of
+        'default_parent_id'
+        context key, or None if it cannot be resolved to a single
+        account.analytic.account
+        """
+        context = self.env.context or {}
+        if type(context.get('default_parent_id')) == int:
+            return context['default_parent_id']
+        return None
+
+    @api.multi
+    @api.depends('parent_id', 'analytic_account_id',
+                 'analytic_account_id.complete_wbs_code',
+                 'analytic_account_id.code')
+    def _compute_analytic_complete_wbs_code(self):
+        for project in self:
+            project.c_wbs_code = \
+                project.analytic_account_id.complete_wbs_code
+
+    project_child_complete_ids = fields.Many2many(
+        comodel_name='project.project',
+        string="Project Hierarchy",
+        compute="_compute_child"
+    )
+    c_wbs_code = fields.Char(
+        compute="_compute_analytic_complete_wbs_code",
+        string='WBS Code',
+        readonly=True,
+        store=True
+    )
+    account_class = fields.Selection(
+        related='analytic_account_id.account_class',
+        store=True,
+        default='project',
+    )
+
+    @api.multi
+    def copy(self, default=None):
+        if default is None:
+            default = {}
+        default['code'] = self.env['ir.sequence'].next_by_code(
+            'account.analytic.account.code')
+        return super(Project, self).copy(default)
+
+    @api.multi
+    def action_open_child_view(self, module, act_window):
+        """
+        :return dict: dictionary value for created view
+        """
+        res = self.env['ir.actions.act_window'].for_xml_id(module, act_window)
+        domain = []
+        project_ids = []
+        for project in self:
+            child_project_ids = self.search(
```

Comment:
@aheficent IMO I think here create one general method, because this logic use many time.